ResearchKit/Common/ORKTaskViewController.m

```diff
@@ -893,9 +893,18 @@ - (void)showViewController:(ORKStepViewController *)viewController goForward:(BO
         }
     }
     
-    if (step.identifier && ![_managedStepIdentifiers.lastObject isEqualToString:step.identifier]) {
-        [_managedStepIdentifiers addObject:step.identifier];
+    NSString *stepIdentifier = step.identifier;
+    if (stepIdentifier) {
+        NSUInteger indexOfStep = [_managedStepIdentifiers indexOfObject:stepIdentifier];
+        if (indexOfStep == NSNotFound) {
+            [_managedStepIdentifiers addObject:stepIdentifier];
+        } else {
+            // If the step being shown has been visited before, remove all the managed step identifiers visited after it
+            NSUInteger numberManagedOfStepIdentifiers = _managedStepIdentifiers.count;
+            [_managedStepIdentifiers removeObjectsInRange:NSMakeRange(indexOfStep + 1, numberManagedOfStepIdentifiers - (indexOfStep + 1))];
```

Comment:
Maybe the code itself is not that obvious.
Currently we use `push`(when user go forward) and `pop`(when user go backward) to maintain the history stack(`_managedStepIdentifiers`). That's why the code is reside in two places.

And I finally figure it out, your solution is trying to consolidate the code into one place instead of two.

It would works identical with the old code when there is no `designed loop` in a task.

But if there is one a designed loop, for example: 

1.  First visit `step1` 
2. Go forward to `step2`  
3. Go forward to `step3` 
4. Go forward to `step2`.

With your code the final results will only include `step1` and `step2`, and `step3` is omitted, which is incorrect.

Let me know what do you think.